cobra/flux_analysis/find_active_reactions.py

```diff
@@ -0,0 +1,434 @@
+        v_j + \varepsilon z^{+}_j >=  \varepsilon  for j \in J with LB_j >= 0
+        v_j - \varepsilon z^{-}_j <= -\varepsilon  for j \in J with UB_j <= 0
+        v_j + Mz^{+}_j >=  \varepsilon  for j \in J with LB_j < 0 and UB_j > 0
+        v_j - Mz^{-}_j <= -\varepsilon  for j \in J with LB_j < 0 and UB_j > 0
+        v_j \in \mathbb{R}
+        z^{+}_j \in \mathbb{R} for j \in J with LB_j >= 0
+        z^{-}_j \in \mathbb{R} for j \in J with UB_j <= 0
+        z^{+}_j, z^{-}_j \in {0,1} for j \in J with LB_j < 0 and UB_j > 0
+
+    LP version:
+    Solve a number of versions of the LP relaxation of the above problem
+    as follows (cumulative changes in each step):
+    1. Fix all :math:`z^{+}_j, z^{-}_j = 1` for all reversible reactions.
+       Solve the LP to find all active irreversible reactions and
+       some active reversible reactions.
+    2. Fix all :math:`z^{+}_j, z^{-}_j = 1` for all irreversible reactions.
+       Unfix :math:`z^{+}_j` for the reversible reactions not yet found active.
+       Solve the LP to find some active reversible reactions
+    3. Fix all math:`z^{+}_j`. Unfix :math:`z^{-}_j` for reversible reactions
+       not yet found active. Solve the LP to find some active rev. reactions
+    4. Add a randomly weighted min. flux constraint:
+       \sum_{j \in J^{rev,not\: active}}{w_j v_j} >= \varepsilon
+       Solve and update the set of reversible reactions not yet found active
+       (if any) until infeasibility
+    5. Change the sense and R.H.S. of the min. flux constraint in Step 4 to
+       '<= -\varepsilon'. Solve and update until infeasibility
+
+    References
+    ----------
+    Chan, S. H., Wang, L., Dash, S., & Maranas, C. D. (2018). Accelerating flux
+    balance calculations in genome-scale metabolic models by localizing the
+    application of loopless constraints. Bioinformatics, 34(24), 4248-4255.
+
+    """
+
+    solve = solve.lower()
+    if solve not in ["lp", "milp", "fastsnp"]:
+        raise ValueError("Parameter solve must be 'lp', 'milp' or 'fastSNP'.")
+
+    elif solve == "fastsnp":
+        N = fastSNP(model, bigM=bigM, zero_cutoff=zero_cutoff)
+        return [model.reactions[j].id for j in range(len(model.reactions))
+                if N[j, :].any()]
+
+    max_bound = max([max(abs(r.upper_bound), abs(r.lower_bound))
+                     for r in model.reactions])
+    if max_bound < float("inf"):
+        bigM = max(bigM, max_bound)
+
+    eps = normalize_cutoff(model, zero_cutoff)
+    try:
+        # at least 100x feasibility if it is defined
+        feas_tol = model.solver.configuration.tolerances.feasibility
+        eps = max(eps, feas_tol * 100)
+    except:
+        feas_tol = eps
+        LOGGER.debug("Feasibility tolerance not defined")
+
+    if solve == "milp":
+        try:
+            # ensure bigM*z << eps at integrality tolerance limit
+            eps = max(eps, model.solver.configuration.tolerances.integrality *
```

Comment:
Thanks. But here it is important for `eps` to be at least 10x larger then the integrality tolerance if solving MILP. Otherwise there could be integrality violation to satisfy the constraint so that some reactions do not necessarily have flux.
And I was not sure if it is always defined for every solver (i.e. `model.configurationas.tolerances.integrality` is always an attribute), so I added `try`/`except` here hoping to see if that might help pass the test. But it was probably not the issue and is not needed, correct? Is `model.tolerance` not less than integrality tolerance if it is used? Any suggestion? Just want to make sure it is large enough when solving MILP and when integrality is used.

Reply:
Ah, I see. We have simplified the tolerance interface in cobrapy. So setting the `model.tolerance` attempts to set all available tolerances given by the solver. Thus it is also expected that the `model.tolerance` value is the same one for integrality, feasiblity, etc. So yes, using `model.tolerance` _should_ do the right thing and also not raise any exceptions.